src/write_kosak.c

```
@@ -0,0 +1,2159 @@
+    return -1;
+  }
+  size_t current_size = c_avl_size(tree_to_flush->tree);
+  while (current_size > 0) {
+    char *json = wg_flush_formatSomeOfTreeAsJSON(server_ctx, tree_to_flush);
+    if (json == NULL) {
+      ERROR("write_kosak: Error formatting tree as JSON");
+      return -1;
+    }
+    kosatron_temp_dump(auth_header, json);
+    // A successful response is the empty string. An unsuccessful response is
+    // a detailed error message from Monarch.
+    char response[2048];
+    if (wg_curl_get_or_post(
+        response, sizeof(response),
+        url, auth_header, JSON_CONTENT_TYPE_HEADER, json) != 0) {
+      ERROR("write_kosak: Error talking to the endpoint");
+      return -1;
+    }
+    ERROR("write_kosak: response from endpoint was %s", response);
+    sfree(json);
+    size_t new_size = c_avl_size(tree_to_flush->tree);
+    if (new_size == current_size) {
+      ERROR("write_kosak: Failed to make progress flushing tree.");
+      return -1;
+    }
+    current_size = new_size;
+  }
+  return 0;
+}
+
+// Converts the data in the tree into a InsertCollectdTimeseriesPointsRequest
+// message (formatted in JSON format). Returns the result in a buffer owned by
+// the caller. In the event of an error, returns NULL. This method is not
+// guaranteed to empty the tree. The caller should call the method multiple
+// times until the tree is empty. Caller owns the tree.
+static char *wg_flush_formatSomeOfTreeAsJSON(const server_ctx_t *server_ctx,
+                                             tree_ctx_t *tree_ctx) {
+  size_t size = JSON_SOFT_TARGET_SIZE + JSON_LOW_WATER_MARK;
+  char *buffer_start = malloc(size);
+  if (buffer_start == NULL) {
+    ERROR("write_kosak: Couldn't allocate %zd bytes for buffer", size);
+    return NULL;
+  }
+
+  char *buffer = buffer_start;
+  wg_request_CreateCollectdTimeseriesPointsRequest(
+      &buffer, &size, JSON_LOW_WATER_MARK, server_ctx, tree_ctx);
+
+  if (size < 2) {
```

What happens (in the future) when we send a plugin/metric that the api doesn't know about?  Does it silently ignore it?